### PR TITLE
Fix mixed module swc compilation for app router

### LINKED ISSUE #57962: Cannot import from `.mjs` file with locally scoped `module.exports = ` code.
### Link to the code that reproduces this issue

https://codesandbox.io/p/sandbox/modern-currying-yz577d

### To Reproduce

1. Open https://codesandbox.io/p/sandbox/modern-currying-yz577d above
2. Notice the error: "Attempted import error: 'bar' is not exported from './file.mjs' (imported as 'bar')."
3. Comment out unused `module.exports = ` line
4. Notice it works
5. Add the line back, and rename the file to `file.js` (from `file.mjs`)
6. Notice it works

### Current vs. Expected behavior

I would expect that the compiler would ignore `module.exports = {}` where `module` is a local variable.

The above code is similar to [code generated by esbuild](https://github.com/evanw/esbuild/blob/4e11b50fe3178ed0a78c077df78788d66304d379/internal/linker/linker.go#L2397):

```js
var require_entities = __commonJS({
  '../../node_modules/entities/lib/maps/entities.json'(exports, module) {


  module.exports = // internal exports
});
```

### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

```bash
Operating System:
  Platform: darwin
  Arch: arm64
  Version: Darwin Kernel Version 22.6.0: Wed Jul  5 22:22:05 PDT 2023; root:xnu-8796.141.3~6/RELEASE_ARM64_T6000
Binaries:
  Node: 18.18.2
  npm: 9.8.1
  Yarn: 1.22.18
  pnpm: 8.6.2
Relevant Packages:
  next: 14.0.2-canary.8
  eslint-config-next: N/A
  react: 18.2.0
  react-dom: 18.2.0
  typescript: 4.9.4
Next.js Config:
  output: N/A
```

### Which area(s) are affected? (Select all that apply)

App Router

### Additional context

I reproduced locally on canary but the code sandbox repro template is fixed to node 16 so I wasn't sure how to upgrade. Still the behaviour seems the same.

### LINKED ISSUE #57964: Attempted import error: 'hls.js' does not contain a default export
### Link to the code that reproduces this issue

https://github.com/laugharn/next-mux-repro

### To Reproduce

1. Start the application in development (pnpm run dev)
2. Go to http://localhost:3000
3. See the import in in app/page.jsx causes the compile error

### Current vs. Expected behavior

The export from hls.js should be detected correctly. I'm using the @muxinc library for demo purposes because that's how this emerged in our project, but importing hls.js directly has the same effect.

### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

```bash
Operating System:
  Platform: darwin
  Arch: arm64
  Version: Darwin Kernel Version 22.6.0: Fri Sep 15 13:41:28 PDT 2023; root:xnu-8796.141.3.700.8~1/RELEASE_ARM64_T6020
Binaries:
  Node: 18.18.2
  npm: 9.8.1
  Yarn: N/A
  pnpm: 8.10.0
Relevant Packages:
  next: 14.0.2-canary.9
  eslint-config-next: N/A
  react: 18.2.0
  react-dom: 18.2.0
  typescript: N/A
Next.js Config:
  output: N/A
```

### Which area(s) are affected? (Select all that apply)

App Router

### Additional context

This looks like it was introduced in 14.0.2-canary.1 and my assumption is #57784 specifically.

### LINKED ISSUE #58831: Next.js v14.0.2-14.0.4-canary.11 break imports `@pathfinder-ide/react` package
### Link to the code that reproduces this issue

https://github.com/jakst/nextjs-pathfinder-import-fail

### To Reproduce

1. Start a build with `pnpm build`
2. Observer the following error (same error in dev server as well)

```
Attempted import error: 's' is not exported from './index-95c7a92d.js' (imported as 'ee').
```

3. Downgrade from Next v14.0.3 (same error in v14.0.4-canary.11) to v14.0.1 `pnpm add next@14.0.1`
4. Run `pnpm build` again
5. Observe that the build passed. You can als `pnpm start` and visit the page to see that the IDE is properly instantiated.

### Current vs. Expected behavior

It should be possible to build the application while importing the `@pathfinder-ide/react` repo, like before

### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

```bash
Operating System:
  Platform: darwin
  Arch: arm64
  Version: Darwin Kernel Version 23.1.0: Mon Oct  9 21:32:11 PDT 2023; root:xnu-10002.41.9~7/RELEASE_ARM64_T6030
Binaries:
  Node: 21.2.0
  npm: 10.2.3
  Yarn: 4.0.2
  pnpm: 8.10.5
Relevant Packages:
  next: 14.0.4-canary.11
  eslint-config-next: N/A
  react: 18.2.0
  react-dom: 18.2.0
  typescript: N/A
Next.js Config:
  output: N/A
```

### Which area(s) are affected? (Select all that apply)

App Router, SWC minifier (swcMinify: true), SWC transpilation

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by huozhi**
After we landed #57784 we started to prefer ESM assets of 3rd party packages in app router bundles including both server (RSC, SSR layers) and client. But some of the packages are not 100% correct format for ES Module.

The packages triggered the bundling errors:

e.g. `@pathfinder-ide/react` includes the below code
```
if (typeof module < "u")
  try {
    module.exports = Fd, Object.defineProperty(Fd, "__esModule", { value: !0 }), Fd.default = Fd, Fd.outdent = Fd;
  } catch {
 ```

which makes next-swc compiler think it's a CJS module since we didn't specify the output module type. Then you saw sometimes import a module will result into failed module exports resolving.

In #57962 it's the same case that when you import the mixed type of code into page, it might be compiled to CJS so you lost the ESM exports.

### Solution

To align the preferred module type and compiled result, we explicitly specify the output module type to "esm", then it will still bundle as ESM. Though mixing two types of code is dangerous and will make the module type ambiguous.

Fixes #58831 
Fixes #57964
Fixes #57962

Closes #58572 (incorrect fix)

### CODE REVIEW

**Comment by ijjk** [comment]
## Failing test suites 
Commit: beaaaf64fcc5a589487007cc17d7c507594a1318

`pnpm test-start test/e2e/app-dir/app-external/app-external.test.ts` 
- app dir - external dependency > should support dynamic import for esm

<details>
<summary>Expand output</summary>

  ● app dir - external dependency › should support dynamic import for esm

    expect(received).toBe(expected) // Object.is equality

    Expected: "esm-with-react"
    Received: ""

      218 |     it('should support dynamic import for esm', async () => {
      219 |       const browser = await next.browser('/esm-dynamic')
    > 220 |       expect(await browser.elementByCss('#dynamic-esm-version').text()).toBe(
          |                                                                         ^
      221 |         'esm-with-react'
      222 |       )
      223 |     })

      at Object.toBe (e2e/app-dir/app-external/app-external.test.ts:220:73)

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

**Comment by ijjk** [comment]
# Stats from current PR

<details open>
<summary><strong>Default Build</strong></summary>

<br/>

<details>
<summary><strong>General</strong></summary>

|  | vercel/next.js canary  | vercel/next.js fix/esm-external-dynamic | Change |
| - | - | - | - |
| buildDuration | 10.2s | 10s | N/A |
| buildDurationCached | 5.9s | 5.9s | N/A |
| nodeModulesSize | 199 MB | 199 MB | N/A |
| nextStartRea..uration (ms) | 426ms | 416ms | N/A |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong></summary>

|  | vercel/next.js canary  | vercel/next.js fix/esm-external-dynamic | Change |
| - | - | - | - |
| 199-HASH.js gzip | 28.7 kB | 28.7 kB | N/A |
| 3f784ff6-HASH.js gzip | 53.3 kB | 53.3 kB | ✓ |
| 494.HASH.js gzip | 180 B | 181 B | N/A |
| framework-HASH.js gzip | 45.2 kB | 45.2 kB | ✓ |
| main-app-HASH.js gzip | 241 B | 238 B | N/A |
| main-HASH.js gzip | 31.7 kB | 31.7 kB | N/A |
| webpack-HASH.js gzip | 1.7 kB | 1.7 kB | ✓ |
| Overall change | 100 kB | 100 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | vercel/next.js fix/esm-external-dynamic | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | vercel/next.js fix/esm-external-dynamic | Change |
| - | - | - | - |
| _app-HASH.js gzip | 194 B | 195 B | N/A |
| _error-HASH.js gzip | 182 B | 181 B | N/A |
| amp-HASH.js gzip | 501 B | 503 B | N/A |
| css-HASH.js gzip | 322 B | 323 B | N/A |
| dynamic-HASH.js gzip | 2.5 kB | 2.5 kB | ✓ |
| edge-ssr-HASH.js gzip | 253 B | 255 B | N/A |
| head-HASH.js gzip | 348 B | 347 B | N/A |
| hooks-HASH.js gzip | 369 B | 368 B | N/A |
| image-HASH.js gzip | 4.27 kB | 4.27 kB | N/A |
| index-HASH.js gzip | 256 B | 256 B | ✓ |
| link-HASH.js gzip | 2.61 kB | 2.6 kB | N/A |
| routerDirect..HASH.js gzip | 311 B | 311 B | ✓ |
| script-HASH.js gzip | 384 B | 383 B | N/A |
| withRouter-HASH.js gzip | 307 B | 308 B | N/A |
| 1afbb74e6ecf..834.css gzip | 106 B | 106 B | ✓ |
| Overall change | 3.17 kB | 3.17 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | vercel/next.js fix/esm-external-dynamic | Change |
| - | - | - | - |
| _buildManifest.js gzip | 484 B | 483 B | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | vercel/next.js fix/esm-external-dynamic | Change |
| - | - | - | - |
| index.html gzip | 528 B | 527 B | N/A |
| link.html gzip | 541 B | 541 B | ✓ |
| withRouter.html gzip | 524 B | 523 B | N/A |
| Overall change | 541 B | 541 B | ✓ |

</details>

<details>
<summary><strong>Edge SSR bundle Size</strong></summary>

|  | vercel/next.js canary  | vercel/next.js fix/esm-external-dynamic | Change |
| - | - | - | - |
| edge-ssr.js gzip | 92.4 kB | 92.4 kB | N/A |
| page.js gzip | 145 kB | 145 kB | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Middleware size</strong></summary>

|  | vercel/next.js canary  | vercel/next.js fix/esm-external-dynamic | Change |
| - | - | - | - |
| middleware-b..fest.js gzip | 627 B | 625 B | N/A |
| middleware-r..fest.js gzip | 150 B | 151 B | N/A |
| middleware.js gzip | 35.6 kB | 35.6 kB | N/A |
| edge-runtime..pack.js gzip | 1.92 kB | 1.92 kB | ✓ |
| Overall change | 1.92 kB | 1.92 kB | ✓ |

</details>

<details>
<summary><strong>Next Runtimes</strong></summary>

|  | vercel/next.js canary  | vercel/next.js fix/esm-external-dynamic | Change |
| - | - | - | - |
| app-page-exp...dev.js gzip | 167 kB | 167 kB | ✓ |
| app-page-exp..prod.js gzip | 93.3 kB | 93.3 kB | ✓ |
| app-page-tur..prod.js gzip | 94.1 kB | 94.1 kB | ✓ |
| app-page-tur..prod.js gzip | 88.6 kB | 88.6 kB | ✓ |
| app-page.run...dev.js gzip | 137 kB | 137 kB | ✓ |
| app-page.run..prod.js gzip | 88 kB | 88 kB | ✓ |
| app-route-ex...dev.js gzip | 23.8 kB | 23.8 kB | ✓ |
| app-route-ex..prod.js gzip | 16.4 kB | 16.4 kB | ✓ |
| app-route-tu..prod.js gzip | 16.5 kB | 16.5 kB | ✓ |
| app-route-tu..prod.js gzip | 16 kB | 16 kB | ✓ |
| app-route.ru...dev.js gzip | 23.2 kB | 23.2 kB | ✓ |
| app-route.ru..prod.js gzip | 16 kB | 16 kB | ✓ |
| pages-api-tu..prod.js gzip | 9.37 kB | 9.37 kB | ✓ |
| pages-api.ru...dev.js gzip | 9.64 kB | 9.64 kB | ✓ |
| pages-api.ru..prod.js gzip | 9.37 kB | 9.37 kB | ✓ |
| pages-turbo...prod.js gzip | 21.8 kB | 21.8 kB | ✓ |
| pages.runtim...dev.js gzip | 22.5 kB | 22.5 kB | ✓ |
| pages.runtim..prod.js gzip | 21.8 kB | 21.8 kB | ✓ |
| server.runti..prod.js gzip | 49.1 kB | 49.1 kB | ✓ |
| Overall change | 924 kB | 924 kB | ✓ |

</details>

<details>
<summary><strong>Diff details</strong></summary>

<details>
<summary>Diff for <strong>page.js</strong></summary>

```diff
failed to diff
```
</details>

</details>

</details>
Commit: beaaaf64fcc5a589487007cc17d7c507594a1318

**Comment by lukasIO** [comment]
> But some of the packages are not 100% correct format for ES Module.

could you elaborate on this? Happy to make adjustments for our libraries if that's the case. 

My finding was (reproduced both in https://github.com/vercel/next.js/issues/58592 and https://github.com/vercel/next.js/issues/57962) that the bug is triggered when there's a locally scoped reference to `module.exports`. E.g.

```js
function (module) {
  module.exports = "this should just be a locally scoped var assignment with no side effects"
}
```

makes the detection fail.

**Comment by huozhi** [comment]
@lukasIO one case was what mentioned in #57962 that the mjs is including both `module.exports` and `export const`, I think that's where confused swc to determine the module type. Since the default module type is not set, so swc will determine module type based on the code. It could be detection issue but for app router layers we should explicitly set the module type to ESM

```js
(module) => {
  module.exports = {};
};

export const bar = "baz";
```

**Comment by lukasIO** [comment]
got it, thanks for the insight. 
I agree that it sounds like the SWC module detection is failing to take the scope into account.

File: packages/next/src/build/swc/options.ts [hunk 1]
```diff
@@ -209,6 +209,20 @@ function getStyledComponentsOptions(
...
+Output module type
+
+For app router where server components is enabled, we prefer to bundle es6 modules,
+Use output module es6 to make sure:
+- the esm module is present
+- if the module is mixed syntax, the esm + cjs code are both present
+
+For pages router will remain untouched
+*/
+function getModuleOptions(esm: boolean) {
```
Comment (timneutkens) [inline packages/next/src/build/swc/options.ts:222]:
```suggestion
function getModuleOptions(esm: boolean): { type: 'es6' } | {} {
```

File: packages/next/src/build/swc/options.ts [hunk 2]
```diff
@@ -423,7 +440,7 @@ export function getLoaderSWCOptions({
               type: 'commonjs',
             },
           }
-        : {}),
+        : getModuleOptions(!!esm)),
```
Comment (timneutkens) [inline packages/next/src/build/swc/options.ts:443]: Better to use explicit type like:

```
esm: boolean | undefined 
```

and then add a default (`= false`) when destructuring or in the return than to cast between undefined/boolean.

File: packages/next/src/build/webpack-config.ts [hunk 3]
```diff
@@ -437,17 +437,26 @@ export default async function getBaseWebpackConfig(
...
     serverComponents: true,
     isReactServerLayer: true,
+    esm: true,
   })
   const swcClientLayerLoader = getSwcLoader({
     serverComponents: true,
     isReactServerLayer: false,
+    esm: true,
+  })
+  // Default swc loaders for pages, not prefer ESM, set `esm` to false
```
Comment (timneutkens) [inline packages/next/src/build/webpack-config.ts:451]:
```suggestion
  // Default swc loaders for pages doesn't prefer ESM.
```

File: packages/next/src/build/handle-externals.ts [hunk 4]
```diff
@@ -28,8 +28,9 @@ export function isResourceInPackages(
   resource: string,
   packageNames?: string[],
```
Comment (timneutkens) [inline packages/next/src/build/handle-externals.ts:29]:
```suggestion
  packageNames: string[],
```

File: packages/next/src/build/handle-externals.ts [hunk 5]
```diff
@@ -28,8 +28,9 @@ export function isResourceInPackages(
...
   packageDirMapping?: Map<string, string>
-) {
-  return packageNames?.some((p: string) =>
+): boolean {
+  if (!packageNames) return false
```
[the remark this answers is not in the document]
Comment (timneutkens) [reply packages/next/src/build/handle-externals.ts:32]: Had a look at the types and it seems you can just change `packageNames` to not accept undefined actually.

```suggestion
```

**Review by timneutkens — APPROVED** [review]

**Pull request merged**